packages/fork-choice/src/protoArray/protoArray.ts

```diff
@@ -741,23 +733,66 @@ export class ProtoArray {
     // If block is from a previous epoch, filter using unrealized justification & finalization information
     // If block is from the current epoch, filter using the head state's justification & finalization information
     const isFromPrevEpoch = computeEpochAtSlot(node.slot) < currentEpoch;
-    const nodeJustifiedEpoch = isFromPrevEpoch ? node.unrealizedJustifiedEpoch : node.justifiedEpoch;
-    const nodeJustifiedRoot = isFromPrevEpoch ? node.unrealizedJustifiedRoot : node.justifiedRoot;
-    const nodeFinalizedEpoch = isFromPrevEpoch ? node.unrealizedFinalizedEpoch : node.finalizedEpoch;
-    const nodeFinalizedRoot = isFromPrevEpoch ? node.unrealizedFinalizedRoot : node.finalizedRoot;
-
-    // If previous epoch is justified, pull up all tips to at least the previous epoch
-    if (this.countUnrealizedFull && currentEpoch > GENESIS_EPOCH && this.justifiedEpoch === previousEpoch) {
-      return node.unrealizedJustifiedEpoch >= previousEpoch;
-      // If previous epoch is not justified, pull up only tips from past epochs up to the current epoch
-    } else {
-      const correctJustified =
-        (nodeJustifiedEpoch === this.justifiedEpoch && nodeJustifiedRoot === this.justifiedRoot) ||
-        this.justifiedEpoch === 0;
+    const votingSourceEpoch = isFromPrevEpoch ? node.unrealizedJustifiedEpoch : node.justifiedEpoch;
+
+    // The voting source should be at the same height as the store's justified checkpoint
+    let correctJustified = votingSourceEpoch === this.justifiedEpoch || this.justifiedEpoch === 0;
+
+    // If this is a pulled-up block from the current epoch, also check that
+    // the unrealized justification is higher than the store's justified checkpoint, and
+    // the voting source is not more than two epochs ago.
+    if (!correctJustified && currentEpoch > GENESIS_EPOCH && this.justifiedEpoch === previousEpoch) {
+      correctJustified = node.unrealizedJustifiedEpoch >= previousEpoch && votingSourceEpoch + 2 >= currentEpoch;
+    }
+
+    try {
+      const finalizedSlot = computeStartSlotAtEpoch(this.finalizedEpoch);
       const correctFinalized =
-        (nodeFinalizedEpoch === this.finalizedEpoch && nodeFinalizedRoot === this.finalizedRoot) ||
-        this.finalizedEpoch === 0;
+        this.finalizedRoot === this.getAncestor(node.blockRoot, finalizedSlot) || this.finalizedEpoch === 0;
       return correctJustified && correctFinalized;
+    } catch (_) {
```

Comment:
Can you refactor getAncestor to not throw to remove this blind try catch? Or scope the try catch to affect getAncestor call only

Reply:
`getAncestor()` is designed to throw so I don't want to break any existing flow. I added `getAncestorOrNull()` and use it here instead